[MIOpen] GTests for 2D CK group convolution solvers

Vsevolod1983 opened the pull request:
Tests For 2D Solvers
ConvHipImplicitGemmGroupFwdXdlops
ConvHipImplicitGemmGroupBwdXdlops
ConvHipImplicitGemmGroupWrwXdlops

It looks like Int8 type for Wrw solver is not supported by CK. I switched test off for I8

Note for reviewers: Please double check Test Cases I added. Feel free to suggest if anything is missing.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,196 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
```
I think the small license header without dates should be used on new files:

```
// Copyright © Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier:  MIT
```

BrianHarrisonAMD replied:
Same for other added files as well.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,196 @@
+#define SOLVER_NAME_DEV_APP DISABLED_ConvHipImplicitGemmGroupBwdXdlops
+#else
+#define SOLVER_NAME_DEV_APP ConvHipImplicitGemmGroupBwdXdlops
+#endif
```
Are these needed for the new test?

I seem to remember these were added previously due to issues with how applicability was being checked on gfx1100 (or any ASIC that doesn't have CK support).  Instead of bulk disabling these checks, could we have the workaround skip the check on devices that don't support CK?

BrianHarrisonAMD replied:
Same for other added files as well.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,196 @@
+                 datatype, {{0, 0}, {1, 1}, {1, 1}}},
+        // clang-format on
+    };
+}
```
I think we should have a test with NCHW layout for smoke as well.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,196 @@
+
+        // clang-format on
+    };
+}
```
I think we need checks for NCHW as well here.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,196 @@
+                 datatype, {{0, 0}, {1, 1}, {3, 3}, 2}}, // dilation > 1
+        TestCase{{datatype, miopenTensorNHWC, {1, 64, 24, 48}},
+                 {datatype, miopenTensorNHWC, {96, 16, 1, 1}},
+                 datatype, {{0, 0}, {1, 1}, {1, 1}, 4}}, // some different NCHW and k parameters
```
I think these are grouped conv cases as well right?
It would be nice to mention that in the comments to make it clear that these are covering grouped cases.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupFwdXdlops.cpp:
```diff
@@ -0,0 +1,188 @@
+                 datatype, {{0, 0}, {1, 1}, {2, 2}, 2}}, // dilation > 1
+        TestCase{{datatype, miopenTensorNHWC, {1, 64, 24, 48}},
+                 {datatype, miopenTensorNHWC, {384, 16, 1, 1}},
+                 datatype, {{0, 0}, {1, 1}, {1, 1}, 4}}, // some different NCHW and k parameters
```
Can probably reduce the sizes of these to improve runtime a bit, and also need NCHW layout coverage.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupFwdXdlops.cpp:
```diff
@@ -0,0 +1,188 @@
+        // clang-format off
+        TestCase{{datatype, miopenTensorNHWC, {1, 64, 8, 8}},
+                 {datatype, miopenTensorNHWC, {96, 64, 1, 1}},
+                 datatype, {{0, 0}, {1, 1}, {1, 1}}},
```
Would be good to add one smoke for NCHW as well I think.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupWrwXdlops.cpp:
```diff
@@ -0,0 +1,171 @@
+
+        // clang-format on
+    };
+}
```
Nit:
Could we use the same style as the other cases when declaring the TestCase?
Not a huge deal, but makes it harder to know what this is testing compared to the others.

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupWrwXdlops.cpp:
```diff
@@ -0,0 +1,171 @@
+    return std::vector{
+        // clang-format off
+
+        TestCase{{8, 8, 3, 3}, {8, 8, 1, 1}, {0, 0}, {1, 1}, {1, 1}, 1, datatype},
```
Takes some digging, but I think this defaults to NCHW, and we are missing a smoke for NHWC in this suite.

https://github.com/ROCm/rocm-libraries/blob/012ea11316db4e703d8ea3059fad4f2134cb1e32/projects/miopen/test/gtest/unit_TensorDescriptor.hpp#L81-L88

https://github.com/ROCm/rocm-libraries/blob/012ea11316db4e703d8ea3059fad4f2134cb1e32/projects/miopen/src/tensor.cpp#L272-L279

https://github.com/ROCm/rocm-libraries/blob/012ea11316db4e703d8ea3059fad4f2134cb1e32/projects/miopen/src/tensor.cpp#L91-L99

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupWrwXdlops.cpp:
```diff
@@ -0,0 +1,171 @@
+
+        TestCase{{datatype, miopenTensorNHWC, {1, 64, 8, 8}},
+                 {datatype, miopenTensorNHWC, {96, 16, 1, 1}},
+                 datatype, {{0, 0}, {2, 2}, {1, 1}, 4}}, // stride > 1
```
Should have NCHW as well here, but these sizes look good!

Vsevolod1983 commented:
Hi @BrianHarrisonAMD 
I refactored all three tests in order to be able to run all layouts we want using: the following style: testing::Values(miopenTensorNHWC, miopenTensorNCHW)

Currently, we dont have such functionality, that's why I introduced new class - UnitTestConvSolverGroupXDlops which I'mgoing to use in all 6 tests (2D and 3D)

BrianHarrisonAMD commented on projects/miopen/test/gtest/unit_conv_solver_ConvHipImplicitGemmGroupBwdXdlops.cpp:
```diff
@@ -0,0 +1,163 @@
+
+#include "unit_conv_solver_group_xdlops.hpp"
+
+#define SOLVER_NAME_DEV_APP ConvHipImplicitGemmGroupBwdXdlops
```
Can this be removed now?

(Same for other files)
